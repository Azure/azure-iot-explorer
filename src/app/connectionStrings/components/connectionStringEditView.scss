/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
 .connection-string-edit-header {
    margin-left: 20px;
    margin-right: 20px;
 }

 .connection-string-edit-body {
    margin-right: 20px;
    margin-left: 20px;
    .details {
       margin-top: 10px;
    }
 }

 .connection-string-edit-footer {
    display: flex;
    margin-left: 20px;
    margin-right: 20px;
    margin-bottom: 10px;
    button {
        margin-right: 10px;
    }
}

