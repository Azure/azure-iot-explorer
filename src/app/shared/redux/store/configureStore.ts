/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { applyMiddleware, compose, createStore } from 'redux';
import { createLogger } from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import { fromJS } from 'immutable';
import reducer from '../reducer';
import rootSaga from '../../../devices/saga';
import { StateInterface } from '../state';
import { appConfig } from './../../../../appConfig/appConfig';

export default () => {
    const middlewares = [];

    const stateTransformer = (state: StateInterface) => {
        const s = fromJS(state);
        return s.toJS();
    };

    if (appConfig.developmentMode) {
        const logger = createLogger({stateTransformer});
        middlewares.push(logger);
    }

    const sagaMiddleware = createSagaMiddleware();
    middlewares.push(sagaMiddleware);

    // tslint:disable-next-line:no-any
    const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    const store = createStore(
        reducer,
        composeEnhancers(applyMiddleware(...middlewares)),
    );
    // some system (not project) configurations fail to realize that this is valid without an any
    // tslint:disable-next-line:no-any
    sagaMiddleware.run(rootSaga as any);

    return store;
};
