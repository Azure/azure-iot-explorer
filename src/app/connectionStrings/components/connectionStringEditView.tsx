/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Panel, PanelType } from 'office-ui-fabric-react/lib/Panel';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { DefaultButton, PrimaryButton } from 'office-ui-fabric-react/lib/Button';
import { ConnectionStringProperties } from './connectionStringProperties';
import { getConnectionInfoFromConnectionString } from '../../api/shared/utils';
import { generateConnectionStringValidationError } from '../../shared/utils/hubConnectionStringHelper';
import { IoTHubConnectionSettings } from '../../api/services/devicesService';
import { useLocalizationContext } from '../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../localization/resourceKeys';
import './connectionStringEditView.scss';

const LINES_FOR_CONNECTION = 8;

export interface ConnectionStringEditViewProps {
    connectionStringUnderEdit?: string;
    connectionStrings: string[];
    onDismiss(): void;
    onCommit(newConnectionString: string): void;
}

export const ConnectionStringEditView: React.FC<ConnectionStringEditViewProps> = props => {
    const {connectionStringUnderEdit, connectionStrings, onDismiss, onCommit} = props;
    const [connectionString, setConnectionString] = React.useState<string>(connectionStringUnderEdit);
    const [connectionStringValidationKey, setConnectionStringValidationKey] = React.useState<string>(undefined);
    const [connectionSettings, setConnectionSettings] = React.useState<IoTHubConnectionSettings>(undefined);
    const { t } = useLocalizationContext();

    React.useEffect(() => {
        if (connectionString) {
            validateConnectionString(connectionString);
        }
    }, []); // tslint:disable-line:align

    const onConnectionStringChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
         setConnectionString(newValue);
         validateConnectionString(newValue);
    };

    const onCommitClick = () => {
        onCommit(connectionString);
    };

    const onDismissClick = () => {
        onDismiss();
    };

    const validateConnectionString = (updatedConnectionString: string) => {
        let validationKey = generateConnectionStringValidationError(updatedConnectionString) || '';

        if (!validationKey) {
            const extractedConnectionSettings = getConnectionInfoFromConnectionString(updatedConnectionString);
            setConnectionSettings(extractedConnectionSettings);
        }

        // check for duplicates and validate || after setting values (so properties display)
        validationKey = (connectionStrings.indexOf(updatedConnectionString) >= 0  && updatedConnectionString !== connectionStringUnderEdit) ?
        ResourceKeys.connectionStrings.editConnection.validations.duplicate :
        validationKey;

        setConnectionStringValidationKey(validationKey);
    };

    const showProperties = (): boolean => {
        if (!connectionSettings) {
            return false;
        }

        if (connectionStringValidationKey && connectionStringValidationKey !== ResourceKeys.connectionStrings.editConnection.validations.duplicate) {
            return false;
        }

        return true;
    };

    const renderHeader = (): JSX.Element => {
        return (
            <h2 className="connection-string-edit-header">
                {connectionStringUnderEdit ?
                    t(ResourceKeys.connectionStrings.editConnection.title.edit) :
                    t(ResourceKeys.connectionStrings.editConnection.title.add)
                }
            </h2>
        );
    };

    const renderFooter = (): JSX.Element => {
        return (
            <div className="connection-string-edit-footer">
                <PrimaryButton
                   text={t(ResourceKeys.connectionStrings.editConnection.save.label)}
                   ariaLabel={t(ResourceKeys.connectionStrings.editConnection.save.ariaLabel)}
                   onClick={onCommitClick}
                   disabled={connectionStringValidationKey !== ''}
                />
                <DefaultButton
                   text={t(ResourceKeys.connectionStrings.editConnection.cancel.label)}
                   ariaLabel={connectionStringUnderEdit ?
                        t(ResourceKeys.connectionStrings.editConnection.cancel.ariaLabel.edit) :
                        t(ResourceKeys.connectionStrings.editConnection.cancel.ariaLabel.add)
                    }
                   onClick={onDismissClick}
                />
            </div>
        );
    };

    return (
        <Panel
            isOpen={true}
            type={PanelType.medium}
            isBlocking={true}
            isFooterAtBottom={true}
            onRenderHeader={renderHeader}
            onRenderFooter={renderFooter}
            onDismiss={onDismiss}
            closeButtonAriaLabel={
                connectionStringUnderEdit ?
                    t(ResourceKeys.connectionStrings.editConnection.cancel.ariaLabel.edit) :
                    t(ResourceKeys.connectionStrings.editConnection.cancel.ariaLabel.add)
            }
        >
            <div className="connection-string-edit-body">
                <TextField
                    ariaLabel={t(ResourceKeys.connectionStrings.editConnection.editField.ariaLabel)}
                    label={t(ResourceKeys.connectionStrings.editConnection.editField.label)}
                    onChange={onConnectionStringChange}
                    multiline={true}
                    rows={LINES_FOR_CONNECTION}
                    errorMessage={t(connectionStringValidationKey)}
                    value={connectionString}
                    required={true}
                    placeholder={t(ResourceKeys.connectionStrings.editConnection.editField.placeholder)}
                />
                {showProperties() &&
                    <div className="details">
                        <ConnectionStringProperties
                            connectionString={connectionString}
                            hostName={connectionSettings.hostName}
                            sharedAccessKey={connectionSettings.sharedAccessKey}
                            sharedAccessKeyName={connectionSettings.sharedAccessKeyName}
                        />
                    </div>
                }
            </div>
        </Panel>
    );
};
