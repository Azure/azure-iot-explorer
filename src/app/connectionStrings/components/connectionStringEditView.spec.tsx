/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { PrimaryButton, DefaultButton } from 'office-ui-fabric-react/lib/Button';
import { ConnectionStringEditView, ConnectionStringEditViewProps } from './connectionStringEditView';
import configureStore from '../../../app/shared/redux/store/configureStore';

describe('ConnectionStringEdit', () => {
    const connectionString = 'HostName=test.azure-devices-int.net;SharedAccessKeyName=iothubowner;SharedAccessKey=key';

    it('matches snapshot in Add Scenario', () => {
        const props: ConnectionStringEditViewProps = {
            connectionStringUnderEdit: '',
            connectionStrings: [],
            onCommit: jest.fn(),
            onDismiss: jest.fn()
        };

        const wrapper = shallow(<ConnectionStringEditView {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('matches snapshot in Edit Scenario', () => {
        const props: ConnectionStringEditViewProps = {
            connectionStringUnderEdit: 'connectionString',
            connectionStrings: [],
            onCommit: jest.fn(),
            onDismiss: jest.fn()
        };

        const wrapper = shallow(<ConnectionStringEditView {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('matches snapshot in Edit / invalid scenario', () => {
        const props: ConnectionStringEditViewProps = {
            connectionStringUnderEdit: 'connectionString',
            connectionStrings: [],
            onCommit: jest.fn(),
            onDismiss: jest.fn()
        };

        const wrapper = shallow(<ConnectionStringEditView {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('matches snapshot in Edit / valid scenario', () => {
        const props: ConnectionStringEditViewProps = {
            connectionStringUnderEdit: connectionString,
            connectionStrings: [],
            onCommit: jest.fn(),
            onDismiss: jest.fn()
        };

        const wrapper = shallow(<ConnectionStringEditView {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('calls onDismiss when Cancel button clicked', () => {
        const onDismiss = jest.fn();
        const props: ConnectionStringEditViewProps = {
            connectionStringUnderEdit: 'connectionString',
            connectionStrings: [],
            onCommit: jest.fn(),
            onDismiss
        };

        const wrapper = mount(<ConnectionStringEditView {...props}/>);
        wrapper.find(DefaultButton).get(1).props.onClick(undefined);

        expect(onDismiss).toHaveBeenCalled();
    });

    describe('edit scenario', () => {
        it('disables commit when validation fails', () => {
            const props: ConnectionStringEditViewProps = {
                connectionStringUnderEdit: connectionString,
                connectionStrings: [],
                onCommit: jest.fn(),
                onDismiss: jest.fn()
            };

            const wrapper = mount(
                <Provider store={configureStore()}>
                     <ConnectionStringEditView {...props}/>
                </Provider>
           );
            act(() => wrapper.find(TextField).props().onChange(undefined, 'badConnectionString'));
            wrapper.update();

            const disabled = wrapper.find(PrimaryButton).props().disabled;
            expect(disabled).toEqual(true);
        });

        it('disables commit when duplicate validation', () => {
            const props: ConnectionStringEditViewProps = {
                connectionStringUnderEdit: '',
                connectionStrings: [connectionString],
                onCommit: jest.fn(),
                onDismiss: jest.fn()
            };

            const wrapper = mount(
                <Provider store={configureStore()}>
                     <ConnectionStringEditView {...props}/>
                </Provider>
           );
            act(() => wrapper.find(TextField).props().onChange(undefined, connectionString));
            wrapper.update();

            const disabled = wrapper.find(PrimaryButton).props().disabled;
            expect(disabled).toEqual(true);
        });

        it('calls onCommit when validation passes', () => {
            const onCommit = jest.fn();
            const props: ConnectionStringEditViewProps = {
                connectionStringUnderEdit: '',
                connectionStrings: [],
                onCommit,
                onDismiss: jest.fn()
            };

            const wrapper = mount(
                <Provider store={configureStore()}>
                     <ConnectionStringEditView {...props}/>
                </Provider>
           );
            act(() => wrapper.find(TextField).props().onChange(undefined, connectionString));
            wrapper.update();

            const commitButton = wrapper.find(PrimaryButton);
            expect(commitButton.props().disabled).toEqual(false);
            commitButton.props().onClick(undefined);

            expect(onCommit).toHaveBeenCalled();
        });

    });
});
