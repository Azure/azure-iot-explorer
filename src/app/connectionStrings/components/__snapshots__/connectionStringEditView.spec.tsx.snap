// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionStringEdit matches snapshot in Add Scenario 1`] = `
<StyledPanelBase
  closeButtonAriaLabel="connectionStrings.editConnection.cancel.ariaLabel.add"
  isBlocking={true}
  isFooterAtBottom={true}
  isOpen={true}
  onDismiss={[MockFunction]}
  onRenderFooter={[Function]}
  onRenderHeader={[Function]}
  type={3}
>
  <div
    className="connection-string-edit-body"
  >
    <StyledTextFieldBase
      ariaLabel="connectionStrings.editConnection.editField.ariaLabel"
      label="connectionStrings.editConnection.editField.label"
      multiline={true}
      onChange={[Function]}
      placeholder="connectionStrings.editConnection.editField.placeholder"
      required={true}
      rows={8}
      value=""
    />
  </div>
</StyledPanelBase>
`;

exports[`ConnectionStringEdit matches snapshot in Edit / invalid scenario 1`] = `
<StyledPanelBase
  closeButtonAriaLabel="connectionStrings.editConnection.cancel.ariaLabel.edit"
  isBlocking={true}
  isFooterAtBottom={true}
  isOpen={true}
  onDismiss={[MockFunction]}
  onRenderFooter={[Function]}
  onRenderHeader={[Function]}
  type={3}
>
  <div
    className="connection-string-edit-body"
  >
    <StyledTextFieldBase
      ariaLabel="connectionStrings.editConnection.editField.ariaLabel"
      label="connectionStrings.editConnection.editField.label"
      multiline={true}
      onChange={[Function]}
      placeholder="connectionStrings.editConnection.editField.placeholder"
      required={true}
      rows={8}
      value="connectionString"
    />
  </div>
</StyledPanelBase>
`;

exports[`ConnectionStringEdit matches snapshot in Edit / valid scenario 1`] = `
<StyledPanelBase
  closeButtonAriaLabel="connectionStrings.editConnection.cancel.ariaLabel.edit"
  isBlocking={true}
  isFooterAtBottom={true}
  isOpen={true}
  onDismiss={[MockFunction]}
  onRenderFooter={[Function]}
  onRenderHeader={[Function]}
  type={3}
>
  <div
    className="connection-string-edit-body"
  >
    <StyledTextFieldBase
      ariaLabel="connectionStrings.editConnection.editField.ariaLabel"
      label="connectionStrings.editConnection.editField.label"
      multiline={true}
      onChange={[Function]}
      placeholder="connectionStrings.editConnection.editField.placeholder"
      required={true}
      rows={8}
      value="HostName=test.azure-devices-int.net;SharedAccessKeyName=iothubowner;SharedAccessKey=key"
    />
  </div>
</StyledPanelBase>
`;

exports[`ConnectionStringEdit matches snapshot in Edit Scenario 1`] = `
<StyledPanelBase
  closeButtonAriaLabel="connectionStrings.editConnection.cancel.ariaLabel.edit"
  isBlocking={true}
  isFooterAtBottom={true}
  isOpen={true}
  onDismiss={[MockFunction]}
  onRenderFooter={[Function]}
  onRenderHeader={[Function]}
  type={3}
>
  <div
    className="connection-string-edit-body"
  >
    <StyledTextFieldBase
      ariaLabel="connectionStrings.editConnection.editField.ariaLabel"
      label="connectionStrings.editConnection.editField.label"
      multiline={true}
      onChange={[Function]}
      placeholder="connectionStrings.editConnection.editField.placeholder"
      required={true}
      rows={8}
      value="connectionString"
    />
  </div>
</StyledPanelBase>
`;
